cmd/mdatagen/templates/match.go.tmpl

```diff
@@ -0,0 +1,492 @@
+// Code generated by mdatagen. DO NOT EDIT.
+
+package metadata
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"sort"
+	"strings"
+	"unicode"
+)
+
+var errNoMatch = errors.New("specified patterns don't match any names")
+
+func expandPatternMap(patternMap map[string]any, matchNames []string) (map[string]any, error) {
+	patterns := patternBlocks{}
+
+	patternStrBlocks := []patternStringBlock{}
+	for patternStr, data := range patternMap {
+		patternStrBlocks = append(patternStrBlocks, patternStringBlock{
+			ptStr: patternStr,
+			data:  data,
+		})
+	}
+
+	for _, patternStrBlock := range patternStrBlocks {
+		patternBlock, err := patternBlockFromStringBlock(patternStrBlock)
+		if err != nil {
+			return nil, err
+		}
+		patterns.add(patternBlock)
+	}
+
+	expandedMap, err := matchAllPatterns(patterns, matchNames)
+	if err != nil {
+		return nil, err
+	}
+	if len(expandedMap) == 0 {
+		return nil, errNoMatch
+	}
+
+	return expandedMap, nil
+}
+
+type groupType int
+
+const (
+	groupWildcard groupType = iota
+	groupMultimatch
+	groupIdentifier
+)
+
+type group struct {
+	typ    groupType
+	values []string
+}
+
+type pattern struct {
+	originalStr string
+	groups      []group
+}
+
+func newPatternFromString(str string) (*pattern, error) {
+	p := &parser{str: str}
+	return p.parse()
+}
+
+type parser struct {
+	str string
+	idx int
+}
+
+func (p *parser) parse() (*pattern, error) {
+	// parse:
+	pt := &pattern{originalStr: p.str}
+	var parseErr error
+parseLoop:
+	for !p.isFinished() {
+		curr, err := p.current()
+		if err != nil {
+			break
+		}
+		switch curr {
+		case '*':
+			g, err := p.parseWildcard()
+			if err != nil {
+				parseErr = err
+				break parseLoop
+			}
+			pt.groups = append(pt.groups, g)
+		case '{':
+			g, err := p.parseMultimatch()
+			if err != nil {
+				parseErr = err
+				break parseLoop
+			}
+			pt.groups = append(pt.groups, g)
+		default:
+			identifier, err := p.parseIdentifier()
+			if err != nil {
+				parseErr = err
+				break parseLoop
+			}
+
+			pt.groups = append(
+				pt.groups,
+				group{typ: groupIdentifier, values: []string{identifier}},
+			)
+		}
+
+		// After a valid group, only a '.' and another group
+		// or end of input are valid.
+		if !p.isFinished() {
+			curr, _ := p.current()
+			if curr != '.' {
```

Comment:
If we allow a group as a whole to be either an identifier, a group, or a wildcard, can we use `strings.Split(pattern, '.')` and just read each part instead of having this loop? I think it should simplify the code.

I like the approach you took. However, even if we do the parsing ourselves, I'd like to keep the code simpler if possible.

Reply:
I'll give that a shot. When I first tried that I was having trouble making nice errors messages, but I think I can come up with a way after all. Will give it a try soon

Reply:
I tried this but I found it was becoming equally as complex. The complexity saved by this approach is:
* Shared scanner object, so I'm not needing to pass around indexes for context to different spots
* I can share identifier parsing code between an identifier as a root group, and an identifier as part of a multimatch

How strongly do you feel on this? I can try to make it work if you think it's the way to go.